flatfs.go

```diff
@@ -1217,46 +1094,542 @@ func (fs *Datastore) Close() error {
+// transaction isolation.
+//
+// Performance characteristics:
+// - Put: O(1) with async I/O, returns immediately
+// - Get/Has/GetSize: O(n) where n = number of Put operations in batch
+// - Commit: O(n) file renames
+//
+// IMPORTANT: Batch instances should not be reused after Commit or Discard.
 type flatfsBatch struct {
-	puts    map[datastore.Key][]byte
+	mu      sync.Mutex
+	puts    []datastore.Key            // ordered list for iteration (Commit, Query)
+	putSet  map[datastore.Key]struct{} // O(1) lookup for Get/Has/GetSize
 	deletes map[datastore.Key]struct{}
 
-	ds *Datastore
+	ds      *Datastore
+	tempDir string
+
+	// Async write tracking
+	asyncWrites     sync.WaitGroup
+	asyncMu         sync.Mutex
+	asyncFirstError error
+	asyncPutGate    chan struct{}
 }
 
 func (fs *Datastore) Batch(_ context.Context) (datastore.Batch, error) {
+	// Create a unique temp directory for this batch
+	// Note: Temp files are not sharded (flat structure) for simplicity and speed.
+	// Files are only sharded when renamed to their final destination on Commit.
+	tempDir := filepath.Join(fs.tempPath, fmt.Sprintf("batch-%d-%d", time.Now().UnixNano(), rand.Int63()))
+	if err := os.Mkdir(tempDir, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create batch temp directory: %w", err)
+	}
+
 	return &flatfsBatch{
-		puts:    make(map[datastore.Key][]byte),
-		deletes: make(map[datastore.Key]struct{}),
-		ds:      fs,
+		putSet:       make(map[datastore.Key]struct{}),
+		deletes:      make(map[datastore.Key]struct{}),
+		ds:           fs,
+		tempDir:      tempDir,
+		asyncPutGate: make(chan struct{}, maxConcurrentPuts),
 	}, nil
 }
 
+// Put writes val for key to a temporary file asynchronously and returns immediately.
+//
+// CRITICAL: The caller MUST NOT modify or reuse the val byte slice after calling Put.
+// The buffer is used asynchronously by a background goroutine. Violating this will
+// cause data corruption. This differs from typical Go semantics where buffers can
+// be reused after a function returns.
+//
+// If you need to reuse buffers, copy them before calling Put:
+//
+//	buf := make([]byte, len(data))
+//	copy(buf, data)
+//	batch.Put(ctx, key, buf)
+//
+// Error handling: If an async write fails, the error is captured and returned
+// on the next Put/Delete/Commit or any read operation (fail-fast behavior).
 func (bt *flatfsBatch) Put(ctx context.Context, key datastore.Key, val []byte) error {
 	if !keyIsValid(key) {
 		return fmt.Errorf("when putting '%q': %v", key, ErrInvalidKey)
 	}
-	bt.puts[key] = val
+
+	if err := bt.getAsyncError(); err != nil {
+		// If there was an error from a previous async write, return it fast
+		// This may be useful if, for example, we are out of disk space.
+		return err
+	}
+
+	// Acquire semaphore slot before starting another async put.
+	select {
+	case bt.asyncPutGate <- struct{}{}:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	bt.mu.Lock()
```

Comment:
hm.. if the same key is Put twice in one batch (a file has some duplicate blocks - rare, but possible e.g. with areas filled with zeroes etc), both goroutines write to the same temp file path. 

Unsure what would happen, but its either an error or corrupted block data, so both outcomes are undesired.

Maybe add check if block already exists and exit early?


```suggestion
	bt.mu.Lock()
	
	// Skip duplicate keys to prevent concurrent goroutines from writing to
	// the same temp file. Without this check, two Put calls with the same key
	// would spawn goroutines that race on os.Create/Write, potentially
	// corrupting the file contents. This maintains the old map-based behavior
	// where duplicate keys simply overwrote (last write wins, atomically).
	if _, exists := bt.putSet[key]; exists {
	  bt.mu.Unlock()
	  <-bt.asyncPutGate // Release semaphore slot acquired above
	  return nil
	}
```